app/src/main/java/io/gnosis/safe/notifications/models/PushNotification.kt

```diff
@@ -0,0 +1,94 @@
+
+    data class NewConfirmation(
+        override val safe: Solidity.Address,
+        val safeTxHash: String,
+        val owner: Solidity.Address
+    ) : PushNotification(TYPE) {
+        companion object {
+            const val TYPE = "NEW_CONFIRMATION"
+            fun fromMap(params: Map<String, String>) =
+                NewConfirmation(
+                    params.getOrThrow("address").asEthereumAddress()!!,
+                    params.getOrThrow("safeTxHash"),
+                    params.getOrThrow("owner").asEthereumAddress()!!
+                )
+        }
+    }
+
+    data class ExecutedTransaction(
+        override val safe: Solidity.Address,
+        val safeTxHash: String,
+        val failed: Boolean
+    ) : PushNotification(TYPE) {
+        companion object {
+            const val TYPE = "EXECUTED_MULTISIG_TRANSACTION"
+            fun fromMap(params: Map<String, String>) =
+                ExecutedTransaction(
+                    params.getOrThrow("address").asEthereumAddress()!!,
+                    params.getOrThrow("safeTxHash"),
+                    params.getOrThrow("failed").toBoolean()
+                )
+        }
+    }
+
+    data class IncomingToken(
+        override val safe: Solidity.Address,
+        val txHash: String,
+        val value: BigInteger,
+        val tokenAddress: Solidity.Address,
+        val tokenId: String? // for ERC721 tokens
+    ) : PushNotification(TYPE) {
+        companion object {
+            const val TYPE = "INCOMING_TOKEN"
+            fun fromMap(params: Map<String, String>) =
+                IncomingToken(
+                    params.getOrThrow("address").asEthereumAddress()!!,
+                    params.getOrThrow("txHash"),
+                    params.getOrThrow("value").toBigInteger(),
+                    params.getOrThrow("tokenAddress").asEthereumAddress()!!,
+                    params["tokenId"]
+                )
+        }
+    }
+
+    data class IncomingEther(
+        override val safe: Solidity.Address,
+        val txHash: String,
+        val value: BigInteger
+    ) : PushNotification(TYPE) {
+        companion object {
+            const val TYPE = "INCOMING_ETHER"
+            fun fromMap(params: Map<String, String>) =
+                IncomingEther(
+                    params.getOrThrow("address").asEthereumAddress()!!,
+                    params.getOrThrow("txHash"),
+                    params.getOrThrow("value").toBigInteger()
+                )
+        }
+    }
+
+    companion object {
+        fun fromMap(params: Map<String, String>) =
+            when (params["type"]) {
+                "NEW_CONFIRMATION" -> NewConfirmation.fromMap(params)
+                "EXECUTED_MULTISIG_TRANSACTION" -> ExecutedTransaction.fromMap(params)
+                "INCOMING_TOKEN" -> IncomingToken.fromMap(params)
+                "INCOMING_ETHER" -> IncomingEther.fromMap(params)
+                else -> throw IllegalArgumentException("Unknown push type")
```

Comment:
2 cents: I hope this won't crash the app, as it may hypothetically receive other types of pushes in the future which should be just ignored

Reply:
its catched in `onMessageReceived(message: RemoteMessage)` in firebase service and it will fail silently then

Reply:
Yeah had the same concern, same goes for the `getOrThrow`